### PR TITLE
tests: verify that simultaneous refresh of kernel and base triggers a single reboot only

### PR DESCRIPTION
**Pull request by bboozzoo**
Verify the single reboot feature. Sadly I was not able to make the test work correctly for core20 as the pc-kernel snap is repacked when preparing the UC20 system. Nonetheless, I'll keep on trying to have it work on uc20 too.

### CODE REVIEW

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for this test, a couple little nitpicky things

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        test -n "$(cat refresh-change-id)"
+        change_id="$(cat refresh-change-id)"
+        # wait until we observe reboots
+        # shellcheck disable=SC2016
+        retry -n 100 --wait 5 sh -c 'test "$(wc -l < /tmp/mock-shutdown.calls)" -gt "1"'
+        # stop snapd now to avoid snapd waiting for too long and deciding to
+        # error out assuming a rollback across reboot
+        systemctl stop snapd.service snapd.socket
+
+        # both link snaps should be done now, we could not use use snap change
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot/task.yaml:69]: thanks for spotting this, the comment and workaround was there before I added the bit stopping snapd, but then forgot to update the comment

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        # as snapd is busy retrying the auto-connect task and does not respond to
+        # API requests, work around that by inspecting the state directly
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot/task.yaml:71]: thanks for spotting this, the comment and workaround was there before I added the bit stopping snapd, but then forgot to update the comment

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        core_snap=core18
+    fi
+
+    if [ "$SPREAD_REBOOT" = 0 ]; then
+        init_fake_refreshes "$BLOB_DIR" pc-kernel
+        init_fake_refreshes "$BLOB_DIR" "$core_snap"
+
+        # taken from transition_to_recover_mode()
+        cp /bin/systemctl /tmp/orig-systemctl
+        mount -o bind "$TESTSLIB/mock-shutdown" /bin/systemctl
```
Comment (stolowski) [inline tests/core/kernel-and-base-single-reboot/task.yaml:56]: Should we undo this in restore section?

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        fi
+
+        # restore shutdown so that spread can reboot the host
+        tests.cleanup pop
+
+        REBOOT
+    elif [ "$SPREAD_REBOOT" = 1 ]; then
+        change_id="$(cat refresh-change-id)"
+        # XXX: is this sufficiently robust?
+        snap watch "$change_id" || true
```
Comment (stolowski) [inline tests/core/kernel-and-base-single-reboot/task.yaml:104]: Will it time-out in a reasonable time if not ready?

Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot/task.yaml:104]: I think so. I mean, that's what we do in all other tests, so unless it's fundamentally broken we should hit the test timeout eventually.

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        snap changes | MATCH "$change_id\s+(Done|Error)"
+        # we expect re-refresh to fail since the tests uses a fake store
+        snap change "$change_id" > tasks.done
+        MATCH '^Error .* Handling re-refresh' < tasks.done
+        # no other errors
+        grep -v 'Handling re-refresh' < tasks.done | NOMATCH '^Error'
+        # nothing was undone
+        grep -v 'Handling re-refresh' < tasks.done | NOMATCH '^Undone'
```
Comment (stolowski) [inline tests/core/kernel-and-base-single-reboot/task.yaml:112]: Should we in addition have a sanity check that revisions of the snaps are as expected?

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        # the change is complete
+        # XXX systemctl exits with non-0 when in degraded state
+        (systemctl is-system-running || true) | MATCH '(running|degraded)'
+
+        # now we need to revert both snaps for restore to behave properly, start
+        # with the kernel
+        snap revert pc-kernel --revision "$(cat pc-kernel.rev)"
+        REBOOT
+    elif [ "$SPREAD_REBOOT" = 2 ]; then
+        snap watch --last=revert\?
```
Comment (stolowski) [inline tests/core/kernel-and-base-single-reboot/task.yaml:125]: Nice trick with `?`, I had no idea we supported this...

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        # now the base
+        snap revert "$core_snap" --revision "$(cat core.rev)"
+        REBOOT
+    elif [ "$SPREAD_REBOOT" = 3 ]; then
+        snap watch --last=revert\?
+        # we're done
+    fi
```
Comment (stolowski) [inline tests/core/kernel-and-base-single-reboot/task.yaml:132]: maybe: else - fail with error?

**Review by stolowski — APPROVED** [review]
Looks good, a couple of minor suggestions.

File: tests/core/kernel-and-base-single-reboot/task.yaml (hunk 5 above)
Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot/task.yaml:112]: Added

File: tests/core/kernel-and-base-single-reboot/task.yaml (hunk 7 above)
Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot/task.yaml:132]: Added a check

File: tests/core/kernel-and-base-single-reboot/task.yaml [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        echo "This test needs test keys to be trusted"
+        exit
+    fi
+    core_snap=core20
+    if os.query is-core18; then
+        core_snap=core18
+    fi
+
+    readlink /snap/pc-kernel/current > pc-kernel.rev
+    readlink "/snap/$core_snap/current" > core.rev
```
Comment (pedronis) [inline tests/core/kernel-and-base-single-reboot/task.yaml:31]: I'm confused why doing this inside restore?

**Review by pedronis — COMMENTED** [review]
I'm confused by some code, see comment.

We also need a variant were the reboot fallbacks and snapd itself reverts.

File: tests/core/kernel-and-base-single-reboot/task.yaml (hunk 8 above)
Comment (bboozzoo) [reply tests/core/kernel-and-base-single-reboot/task.yaml:25]: This is fixed now

**Pull request merged**